motr/setup.c

```diff
@@ -1545,35 +1545,88 @@ static void cs_part_domain_setup(struct m0_reqh_context *rctx)
 		part_cfg->bpc_location =
 			(char*)cs_storage_partdom_location_gen(sdev->sd_filename);
 		part_cfg->bpc_dom_key = sdev->sd_dev_idx;
-		/** seg0 configuration */
-		part_cfg->bpc_part_mode_seg0 = true;
-		rctx->rc_be_seg0_path = sdev->sd_filename;
-		proposed_chunk_size = sdev->sd_size / def_dev_chunk_count;
-		part_cfg->bpc_chunk_size_in_bits =
-			align_chunk_size(proposed_chunk_size);
-		part_cfg->bpc_total_chunk_count =
-			sdev->sd_size >> part_cfg->bpc_chunk_size_in_bits;
-
-		/** seg1 configuration */
-		m0_bcount_t meta_size = (sdev->sd_size) / 10;
-		part_cfg->bpc_part_mode_seg1 = true;
-		rctx->rc_be_seg_path = sdev->sd_filename;
-		if(rctx->rc_be_seg_size > meta_size){
+		if (rctx->rc_be_seg_path == NULL &&
+		    rctx->rc_be_seg0_path == NULL &&
+		    rctx->rc_be_log_path == NULL ) {
+
+			/** 
+ 			 * Case 1: Single Data and No Meta specified
+ 			 */
+	
+			/** seg0 configuration */	
+			part_cfg->bpc_part_mode_seg0 = true;
+			rctx->rc_be_seg0_path = sdev->sd_filename; 
+			proposed_chunk_size = sdev->sd_size / def_dev_chunk_count;
+			part_cfg->bpc_chunk_size_in_bits =
+				align_chunk_size(proposed_chunk_size);
+			part_cfg->bpc_total_chunk_count =
+				sdev->sd_size >> part_cfg->bpc_chunk_size_in_bits;
+
+			/** seg1 configuration */
+			m0_bcount_t meta_size = (sdev->sd_size) / 10;
+			part_cfg->bpc_part_mode_seg1 = true;
+			rctx->rc_be_seg_path = sdev->sd_filename;
+			if(rctx->rc_be_seg_size > meta_size)
+				rctx->rc_be_seg_size = meta_size;
 			part_cfg->bpc_seg_size_in_chunks =
-				meta_size >> part_cfg->bpc_chunk_size_in_bits;
-			rctx->rc_be_seg_size = meta_size;
-		}
-		else
+				rctx->rc_be_seg_size >> part_cfg->bpc_chunk_size_in_bits;
+
+			/** Log configuration*/
+			part_cfg->bpc_part_mode_log = true;
+			rctx->rc_be_log_path = sdev->sd_filename;
+			if(proposed_chunk_size < def_log_size)
+				part_cfg->bpc_log_size_in_chunks =
+					def_log_size >> part_cfg->bpc_chunk_size_in_bits;
+			else
+				part_cfg->bpc_log_size_in_chunks = 1;
+
+		} else if (rctx->rc_be_seg0_path != NULL &&
+			   rctx->rc_be_seg_path == NULL &&
+			   rctx->rc_be_log_path == NULL ) {
+
+			/** 
+		 	* Case 2: Single Data and Meta specified
+		 	* Case 4: Multiple Data and Meta spcified
+		 	*/
+
+			/** seg0 configuration */	
+			part_cfg->bpc_part_mode_seg0 = true;
+			proposed_chunk_size = rc_be_seg_size / def_dev_chunk_count;
+			part_cfg->bpc_chunk_size_in_bits =
+				align_chunk_size(proposed_chunk_size);
+			part_cfg->bpc_total_chunk_count =
+				rc_be_seg_size >> part_cfg->bpc_chunk_size_in_bits;
+
+			/** seg1 configuration */
+			m0_bcount_t meta_size = (rc_be_seg_size) / 10;
```

Comment:
in this case when only metadata, log and seg0 will be part of part stob; so once chunks are allocated to seg0, log; remaining will go to seg/metadata